extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/filters/CORSFilter.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.undertow.runtime.filters;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+public class CORSFilter implements Filter {
+
+    @ConfigProperty(name = "quarkus.http.cors.origins")
+    public Optional<String> origins;
+    @ConfigProperty(name = "quarkus.http.cors.methods")
+    public Optional<String> methods;
+    @ConfigProperty(name = "quarkus.http.cors.headers")
+    public Optional<String> headers;
+    @ConfigProperty(name = "quarkus.http.cors.exposed-headers")
+    public Optional<String> exposedHeaders;
+
+    private static final String ACCESS_CONTROL_ALLOW_ORIGIN = "Access-Control-Allow-Origin";
+    private static final String ACCESS_CONTROL_ALLOW_CREDENTIALS = "Access-Control-Allow-Credentials";
+    private static final String ACCESS_CONTROL_ALLOW_METHODS = "Access-Control-Allow-Methods";
+    private static final String ACCESS_CONTROL_ALLOW_HEADERS = "Access-Control-Allow-Headers";
+    private static final String ORIGIN = "Origin";
+    private static final String ACCESS_CONTROL_REQUEST_METHOD = "Access-Control-Request-Method";
+    private static final String ACCESS_CONTROL_EXPOSE_HEADERS = "Access-Control-Expose-Headers";
+    private static final String ACCESS_CONTROL_REQUEST_HEADERS = "Access-Control-Request-Headers";
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+        String origin = request.getHeader(ORIGIN);
+        if (origin == null) {
+            chain.doFilter(servletRequest, servletResponse);
+        } else {
+            String requestedMethods = request.getHeader(ACCESS_CONTROL_REQUEST_METHOD);
```

Comment:
Should ALLOW_HEADERS and ALLOW_METHODS be empty if the REQUEST versions aren't set?

I'd always presumed they needed to be set

Reply:
I am really not sure about it. According to the MDN reference https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Preflighted_requests 

There are simple requests that only requires ORIGIN to be set, and there are preflighted requests that will always include HEADERS and METHODS. According to that, this filter will fulfill both scenarios.

But again, I'm not an expert, to be honest I only use this for development and I enable the "*" wildcard for all 😄.